Lokinet cmake/CI fix

dr7ana opened the pull request:
- libquic doesn't need to build oxen-encoding when it's integrated into lokinet
- 004 needs to account for hardware differences when opening streams
- 007 cannot strictly predict ngtcp2 behavior to fine granularity across different architectures

dr7ana committed:
cmake dependency fix for integrating in other oxen projects

dr7ana commented on external/CMakeLists.txt:
```diff
@@ -57,7 +57,9 @@ set(OXEN_LOGGING_SOURCE_ROOT "${PROJECT_SOURCE_DIR}" CACHE INTERNAL "")
 add_subdirectory(oxen-logging)
```
wrap this in if (NOT TARGET ...

dr7ana committed:
004 stream test refactor

- added hardware concurrency check to ensure ad hoc stream creation is OK
- rapid fire stream creation is done through callsafe
- minor change to oxen-logging cmake target check

dr7ana committed:
004 stream test re-refactor

jagerman commented on external/CMakeLists.txt:
```diff
@@ -50,14 +50,18 @@ add_library(ngtcp2::crypto
 # oxen-logging
-if(BUILD_STATIC_DEPS)
-    set(OXEN_LOGGING_FORCE_SUBMODULES ON CACHE INTERNAL "")
+if (NOT TARGET oxen-logging)
```
`oxen-logging` is a semi-internal target for oxen-logging; the public (ALIAS) target is `oxen::logging`

```suggestion
if (NOT TARGET oxen::logging)
```

jagerman commented on external/CMakeLists.txt:
```diff
@@ -50,14 +50,18 @@ add_library(ngtcp2::crypto
+    set(OXEN_LOGGING_SOURCE_ROOT "${PROJECT_SOURCE_DIR}" CACHE INTERNAL "")
+    add_subdirectory(oxen-logging)
 endif()
-set(OXEN_LOGGING_SOURCE_ROOT "${PROJECT_SOURCE_DIR}" CACHE INTERNAL "")
```
If it's already set up we should still append our source root to the current value (with `;` to separate)

jagerman commented on tests/004-streams.cpp:
```diff
@@ -153,24 +153,31 @@ namespace oxen::quic::test
         std::atomic<int> data_check{0};
-        opt::max_streams max_streams{8};
-        std::vector<std::shared_ptr<Stream>> streams{12};
+        int n_threads = (std::thread::hardware_concurrency() < 12) ? std::thread::hardware_concurrency() : 12;
+
+        if (n_threads < 8)
+            SKIP("Current hardware does not support ad hoc creation of > 8 threads!");
```
This isn't necessary; you can still create however many threads you want, even if you don't have the CPUs cores/threads to back them (they just got scheduled across the cores you do have).

jagerman commented on external/CMakeLists.txt:
```diff
@@ -50,14 +50,20 @@ add_library(ngtcp2::crypto
+    set(OXEN_LOGGING_SOURCE_ROOT "${PROJECT_SOURCE_DIR}" CACHE INTERNAL "")
+    add_subdirectory(oxen-logging)
+else()
+    set(OXEN_LOGGING_SOURCE_ROOT "${OXEN_LOGGING_SOURCE_ROOT};${PROJECT_SOURCE_DIR}" CACHE INTERNAL "")
```
Need to investigate whether this appends multiple times if you re-run cmake multiple times from an existing build directory.  (The final value will end up in the build dir's CMakeCache.txt).

jagerman commented on tests/004-streams.cpp:
```diff
@@ -209,47 +212,40 @@ namespace oxen::quic::test
         for (int i = 0; i < 2; ++i)
         {
-            std::thread open_thread([&]() {
+            client_endpoint->call([&]() {
```
A `call` here shouldn't be needed (and, in fact, if we need it, then something is wrong in the API); we ought to be able to just directly call the code here, from the main thread.

jagerman replied:
Neither call nor thread are wanted here.

jagerman commented on tests/004-streams.cpp:
```diff
@@ -209,47 +212,40 @@ namespace oxen::quic::test
+        for (int i = 0; i < n_streams; ++i)
         {
-            std::thread stream_thread([&]() {
+            client_endpoint->call([&]() {
```
`call` shouldn't be needed

jagerman replied:
Neither call nor std::thread are needed here.

jagerman committed:
Fix datagram recombining buffer overuse

`b.data.size()` is the *max* buffer size, but `b.data_size` is how much
we actually used.

jagerman committed:
Add `call_get` to abstract the promise-future-getter pattern

dr7ana committed:
Fixed weirdness

- CI tests rotating buffer
- 004 stream test fixed
- speed test compile flag and atomics sorted

dr7ana committed:
004 stream test refactor 3: the re-re-refactoring

- Okay so we got to the bottom of everything finally. The calls that should be queued are queued, and the user of the API no longer needs to stress about this
